httpx/_dispatch/wsgi.py

```diff
@@ -7,6 +8,15 @@
 from .base import SyncDispatcher
 
 
+def _get_non_empty_chunk(body: typing.Iterable) -> typing.Iterable:
+    """Get a non-empty chunk from body."""
+    body = iter(body)
+    for chunk in body:
+        if chunk:
+            return itertools.chain([chunk], body)
+    return []
```

Comment:
From the PEP:

> In other words, response headers must not be sent until there is actual body data available, or until the application's returned iterable is exhausted. *(The only possible exception to this rule is if the response headers explicitly include a Content-Length of zero.)*

Is it worth explictly checking for the Content-Length to be zero inrrespective of the body also being empty?

I'm not sure if that's something that can ever happen though, not super familiar with WSGI myself.

Reply:
My reading of this is that if a Content-Length of zero is set, then you CAN skip checking for body data, not that you SHOULD skip checking for body data (eg for performance reasons).  It's not something I've ever bothered to put in a WSGI adapter before, though my use cases are always quite lightweight (I mainly use it for testing web apps, which is what I'm doing with httpx).

It would mitigate against a situation where the app returned an infinite empty iterator, as long as it had a Content-Length of zero.  You would still have to take the first value of the iterator though, eg when a generator is being used.

Happy to add it, as it's a tiny amount of code.  It's really your decision over how you want the library to behave.

Reply:
I guess it might be worth it while we're at it but I'd like to hear @tomchristie and @florimondmanca's opinion.

Thanks again @Singletoned

Reply:
Feels like a super narrow edge case to me… I read the paragraph in the PEP as "in case of `Content-Length: 0` then you *don't have to* check for body data" too. So maybe we can leave things as they are for now, in the name of keeping things simple while covering 99.99% of use cases?

Reply:
I guess a broader question here is, in general, do you want httpx to be:

  a) as strict as possible (tightly conforming to specs)
  b) as permissive as possible (eg handle broken responses, etc)
  c) as simple as possible (least code complexity)

Given that it will be used to fetch things on the web, I would guess "b" is the answer, but any of them are valid.  If you do have a clear answer, I'd suggest putting it in the "contributing" docs.

Under either "a" or "b" I'd suggest that this should be added.  It conforms to the spec, and it also allows broken cases to succeed.  Under "c" then it's not really needed.

Reply:
I think for areas like these where neither urllib3 nor Requests support the feature, we're more on a pragmatic variant of c).

If there's something obviously broken we should fix it. But so far we haven't seen any user issues about "my WSGI server returns an infinite stream of empty chunks and HTTPX can't process it", which is why I'd be advocating for "let's fix this only if/when we have actual proof users are encountering it in the wild". :-)